File: src/207.cpp
#include <vector>

class Solution {
public:

    std::vector<vector<int>> graph;
    std::vector<int> visited_arr;


    bool dfs(int course){
        if( visited_arr[course] == 1){
                return false;
            }

            if( visited_arr[course] == 2){
                return true;
            }

            visited_arr[course] = 1;

            for(int i = 0; i < graph[course].size(); ++i){
                if(!dfs(graph[course][i])){
                    return false;
                }
            }

        visited_arr[course] = 2; 
        return true;
    }

    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        graph.resize(numCourses);

        for(int i = 0; i < prerequisites.size(); ++i){
            int a = prerequisites[i][0];
            int b = prerequisites[i][1];
            graph[b].push_back(a);
        }

        visited_arr.resize(numCourses, 0);

        for(int i = 0; i < numCourses; ++i){
            if(visited_arr[i] == 0){
                if(!dfs(i)){
                    return false;
                }
            }
        }

        return true;

    }
};

File: src/207.go
func canFinish(numCourses int, prerequisites [][]int) bool {

	sons := make([][]int, numCourses)
	degree_in := make([]int, numCourses)
	for _, v := range prerequisites {
		sons[v[0]] = append(sons[v[0]], v[1])
		degree_in[v[1]]++
	}

	origin_queue := make([]int, numCourses)
	start, end := 0, 0
	for i, v := range degree_in {
		if v == 0 {
			origin_queue[end] = i
			end++
		}
	}

	for start < end {
		for _, v := range sons[origin_queue[start]] {
			degree_in[v]--
			if degree_in[v] == 0 {
				origin_queue[end] = v
				end++
			}
		}
		start++
	}

	return (end == numCourses)
}